app/queries/facility_analytics_query.rb

```diff
@@ -37,28 +37,20 @@ def registered_patients_by_period
   end
 
   def follow_up_patients_by_period
-    #
-    # this is similar to what the group_by_period query already gives us,
-    # however, groupdate does not allow us to use these "groups" in a where clause
-    # hence, we've replicated its grouping behaviour in order to remove the patients
-    # that were registered prior to the period bucket
-    #
-    date_truncate_string =
-      "(DATE_TRUNC('#{@period}', blood_pressures.recorded_at::timestamptz AT TIME ZONE '#{Groupdate.time_zone || 'Etc/UTC'}'))"
-
     @follow_up_patients_by_period ||=
-      BloodPressure
-        .left_outer_joins(:user)
-        .left_outer_joins(:patient)
-        .joins(:facility)
-        .where(facility: @facility)
-        .where(deleted_at: nil)
-        .group('users.id')
-        .group_by_period(@period, 'blood_pressures.recorded_at')
-        .where("patients.recorded_at < #{date_truncate_string}")
-        .order('users.id')
-        .distinct
-        .count('patients.id')
+      Patient
+        .from(Patient
+                .joins(:blood_pressures)
+                .follow_ups(@period, last: @prev_periods)
+                .group('user_id', 'blood_pressures.patient_id', BloodPressure.date_to_period_sql(@period), 'blood_pressures.recorded_at', 'patients.deleted_at')
+                .where(blood_pressures: { facility: @facility })
+                .select("DISTINCT ON (blood_pressures.patient_id, #{BloodPressure.date_to_period_sql(@period)}) blood_pressures.user_id AS user_id, patients.deleted_at")
+                .select("blood_pressures.recorded_at AS bp_recorded_at")
+                .order('blood_pressures.patient_id', BloodPressure.date_to_period_sql(@period), 'blood_pressures.recorded_at'),
+              'patients')
+        .group('user_id')
+        .group_by_period(:month, 'bp_recorded_at')
+        .count
```

Comment:
This query looks more complex than what was here before. Do we need to change it at all? Could you shed some light on what's going on here?

Reply:
If we don't change it then we'll get that weird _multiple users in the same facility causing overcounted_ follow-ups data issue. I wouldn't want to keep that data issue around even if that page is not used by CVHOs because it's another point of confusion if someone clicks on it and might mistrust the supposedly correct data because now there's two sources of data.

I'll explain how this query works in another comment.

Reply:
The `DISTINCT ON` accounts for the case where a patient visits a facility twice in a period and is seen by different users. Without it, the visit is counted twice and the total increases in comparison to facility follow ups. 

The notable clause being that we're `DISTINCT`ing `ON` the `BloodPressure.date_to_period_sql` **and** the user_id together.

However, we've turned this into a nested subquery using `.from` as a workaround for `ActiveRecord`'s inability to compose a `COUNT` with a `DISTINCT ON` – it ends up with garble like `COUNT DISTINCT DISTINCT ON` which is an invalid query since AR doesn't know how to handle `DISTINCT ON`.

This is a query I'm not happy with either. Though, I also don't want to spend additional time in making it easier to read. I've pushed this as a [comment](https://github.com/simpledotorg/simple-server/pull/888/commits/df832116799d90a81be106b5749927d7d02fbe73) in the code as well for some clarity.

Reply:
Yikes indeed. The comment is very descriptive. 🚀